### PR TITLE
Move CEP CRD into ciliumendpoints.cilium.io API group to allow independent versioning

### PR DESCRIPTION
**Pull request by raybejjani**
_This PR might be best viewed per-commit_

We want to avoid rolling the version number on CNPs when we change the CEP (and vice versa). By creating a separate group we they become independent and users do not need to update their yaml unnecessarily. Note that a similar change for CNPs will occur in https://github.com/cilium/cilium/pull/3164, and this PR should be merged after that one (this one is a simpler rebase).

Move CiliumEndpoints to their own APIGroup allowing us to change their version independent of CiliumNetworkPolicy. The major changes are:
1- Factor out the `pkg/k8s/apis/cilium.io/util` package to `pkg/k8s/apis/util`
2- Extract the CiliumEndpoint CRD setup logic & types into `pkg/k8s/apis/ciliumendpoints.cilium.io`
3- Remove all CiliumEndpoint related stuff from `pkg/k8s/apis/cilium.io`
4- Delete the older CRD to avoid it keeping the kubectl shortnames

Separately, I kept the package name the same as the CRD itself (ciliumendpoints.cilium.io) but since https://github.com/cilium/cilium/pull/3164 uses `cnp.cilium.io` (I think), I can follow that pattern.
The utils.createUpdateCRD function relies on a specific label being on the CRD. It relies on the caller for this right now but could change the CRD itself. I opted to not modify the object since it's a little unexpected.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/3222)

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

File: pkg/k8s/apis/ciliumendpoints.cilium.io/const.go
Comment (aanm) [inline pkg/k8s/apis/ciliumendpoints.cilium.io/const.go]: This is no longer deprecated, we are at `v1` for cilium endpoint

File: pkg/k8s/apis/ciliumendpoints.cilium.io/register.go
Comment (aanm) [inline pkg/k8s/apis/ciliumendpoints.cilium.io/register.go]: This is `ciliumendpoints.cilium.io`

File: pkg/k8s/apis/ciliumendpoints.cilium.io/v2/register.go
Comment (aanm) [inline pkg/k8s/apis/ciliumendpoints.cilium.io/v2/register.go]: Did you test to upgrade a cluster with the old `cep`? I'm just curious to see if the CRD will fail to create due conflicting short names.

File: pkg/k8s/client/clientset/versioned/clientset.go
Comment (aanm) [inline pkg/k8s/client/clientset/versioned/clientset.go]: Why did this message disappeared? You used the `1.10.0-beta.3` generator, right?

File: pkg/k8s/apis/ciliumendpoints.cilium.io/register.go
Comment (aanm) [reply pkg/k8s/apis/ciliumendpoints.cilium.io/register.go]: Also, so we will go with `ciliumendpoints.cilium.io`? for the network policy I was thinking on `cnp.cilium.io` but, for the `CustomResourceDefinitionSingularName` I used `ciliumnetworkpolicy`

Comment (tgraf) [reply pkg/k8s/apis/ciliumendpoints.cilium.io/register.go]: The user is never going to type these things so I suggest to use long names that are clear. Abbreviations  always require to educate the user first.

File: pkg/k8s/client/clientset/versioned/clientset.go
Comment (raybejjani) [reply pkg/k8s/client/clientset/versioned/clientset.go]: probably not, I think I had a 1.9 variant checked out (why that message would change I have no idea)

File: pkg/k8s/apis/ciliumendpoints.cilium.io/register.go
Comment (raybejjani) [reply pkg/k8s/apis/ciliumendpoints.cilium.io/register.go]: I'll try again but I couldn't figure out why the https://github.com/cilium/cilium/pull/3164 PR can use `cnp.cilium.io` without the resulting CRD being `ciliumnetworkpolicies.cnp.cilium.io`. I was aiming to keep the same name we have now. If I made `GroupName = "cilliumendpoints.cilium.io"` the CRD could not be registered because it wants the group field in the CRD to be `CustomResourceDefinitionPluralName + "." + SchemeGroupVersion.Group` which resulted in `ciliumendpoints.ciliumendpoint.cilium.io`. I don't fully understand why this works as it is (or doesn't, for that matter).
If we did intend to namespace all things endpoint related under ciliumendpoint.cilium.io then it might be better to go with the abbreviation for that middle bit.

File: pkg/k8s/apis/ciliumendpoints.cilium.io/v2/register.go
Comment (raybejjani) [reply pkg/k8s/apis/ciliumendpoints.cilium.io/v2/register.go]: Not rigourously. I will do that now. It did return objects, though, so I saw no error (which doesn't mean it worked).

File: pkg/k8s/client/clientset/versioned/clientset.go
Comment (raybejjani) [reply pkg/k8s/client/clientset/versioned/clientset.go]: fixed

**Comment by ianvernon** [comment]
Will this cause any conflicts with the Policy v3 work?

**Comment by raybejjani** [comment]
@ianvernon yes, but mostly rebase conflicts. I did factor out the utils files while the v3 work copies them. Once that is merged I will rebase this PR and make sure everything is still functional.

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by raybejjani** [comment]
@aanm I actually have some changes you should review. Mostly around deleting the old CRD so it doesn't keep the shortnames. I'm testing it now.

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
@aanm I added DeleteCRD in `pkg/k8s/apis/utils/crd.go` and we call it to delete the older `ciliumendpoint.cilium.io` CRD when we add the new `ciliumendpoints.endpoints.cilium.io` crd.

**Comment by raybejjani** [comment]
cc @nebril This will break microscope when merged!

**Comment by raybejjani** [comment]
test-me-please

**Comment by joestringer** [comment]
I set this to WIP since has conflicts with master. Please rebase and return the labels when it's ready for further review.

**Comment by raybejjani** [comment]
@joe Thanks. I should have done that earlier.
@tgraf Is this intended to go into 1.1? (you set the tag) I thought we paused all efforts to sort out versioning on the CRDs, including moving the CEPs into a separate namespace. If so, I can close the PR to minimise confusion.

**Pull request closed without merging**
